flowauth/backend/flowauth/token_management.py

```diff
@@ -134,47 +102,76 @@ def add_token(server):
 
     Notes
     -----
-    Expects json with "name", "expiry" and "claims" keys, where "name" is a string,
-    expiry is a datetime string of the form ""%Y-%m-%dT%H:%M:%S.%fZ" (e.g. 2018-01-01T00:00:00.0Z),
-    and claims is a nested dict of the form {<claim_name>:{<right_name>:<bool>}}.
+    Expects json with "name", and "roles" keys, where "name" is a string,
+    and roles is a list of role names.
 
     Responds with a json object {"token":<token_string>, "id":<token_id>}.
 
     """
-    server = Server.query.filter(Server.id == server).first_or_404()
+    server = Server.query.filter(Server.id == server_id).first_or_404()
     json = request.get_json()
 
     current_app.logger.debug("New token request", request=json)
-
     if "name" not in json:
         raise InvalidUsage("No name.", payload={"bad_field": "name"})
-    expiry = datetime.datetime.strptime(json["expiry"], "%Y-%m-%dT%H:%M:%S.%fZ")
+    if "roles" not in json:
+        raise InvalidUsage("No roles.", payload={"bad_field": "roles"})
+
+    user_roles = db.session.execute(
+        select(Role)
+        .join(User.roles)
+        .filter(User.id == current_user.id)
+        .filter(Role.server_id == server_id)
+    ).all()
+    user_roles = [row[0] for row in user_roles]
+
+    roles = []
+    for requested_role in json["roles"]:
+        try:
+            this_role = next(
+                filter(lambda x: x.name == requested_role["name"], user_roles)
+            )
+        except StopIteration:
+            raise Unauthorized(
+                f"Role '{requested_role['name']}' is not permitted for the current user"
+            )
+        roles.append(
+            {
+                "name": this_role.name,
+                "latest_token_expiry": this_role.latest_token_expiry,
+                "scopes": [s.name for s in this_role.scopes],
+            }
+        )
+
+    expiry = reduce(
+        lambda prev, cur: prev if prev < cur else cur,
+        (role.pop("latest_token_expiry") for role in roles),
+    )
```

Comment:
This finds the earliest "latest expiry" across all applicable roles, which is good, but you also need to take into account the shortest "longest lifetime". (If I'm reading the models correctly, `Role.latest_token_expiry` directly returns the "latest token expiry" setting and is independent of the "longest_token_life_minutes", unlike `User.latest_token_expiry(server)` which calculates latest expiry from a combination of "latest expiry", "longest lifetime" and `datetime.now()`).